From 185abc2a7aca55a6cb39e9ea4ba2fd45a0e19c22 Mon Sep 17 00:00:00 2001
From: Amos Kong <akong@redhat.com>
Date: Wed, 15 Jan 2014 09:58:06 +0100
Subject: [PATCH 37/39] rng-egd: offset the point when repeatedly read from the buffer

RH-Author: Amos Kong <akong@redhat.com>
Message-id: <1389779886-26583-4-git-send-email-akong@redhat.com>
Patchwork-id: 56726
O-Subject: [RHEL-6.6 qemu-kvm PATCH 3/3] rng-egd: offset the point when repeatedly read from the buffer
Bugzilla: 786407
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>
RH-Acked-by: Amit Shah <amit.shah@redhat.com>
RH-Acked-by: Michael S. Tsirkin <mst@redhat.com>

The buffer content might be read out more than once, currently
we just repeatedly read the first data block, buffer offset is
missing.

Cc: qemu-stable@nongnu.org
Signed-off-by: Amos Kong <akong@redhat.com>
Message-id: 1385023371-8198-3-git-send-email-akong@redhat.com
Signed-off-by: Anthony Liguori <aliguori@amazon.com>
(cherry picked from commit 1eb1bd9eafa890f1f4d16ef5cb8b9239a86874d9)
---
 backends/rng-egd.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

Signed-off-by: Miroslav Rezanina <mrezanin@redhat.com>
---
 backends/rng-egd.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/backends/rng-egd.c b/backends/rng-egd.c
index d653641..41c5f9b 100644
--- a/backends/rng-egd.c
+++ b/backends/rng-egd.c
@@ -91,12 +91,14 @@ static int rng_egd_chr_can_read(void *opaque)
 static void rng_egd_chr_read(void *opaque, const uint8_t *buf, int size)
 {
     RngEgd *s = RNG_EGD(opaque);
+    size_t buf_offset = 0;
 
     while (size > 0 && s->requests) {
         RngRequest *req = s->requests->data;
         int len = MIN(size, req->size - req->offset);
 
-        memcpy(req->data + req->offset, buf, len);
+        memcpy(req->data + req->offset, buf + buf_offset, len);
+        buf_offset += len;
         req->offset += len;
         size -= len;
 
-- 
1.7.1

